arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.h

```diff
@@ -39,26 +38,42 @@ namespace arangodb::replication2::replicated_state::document {
 
 struct IDocumentStateShardHandler {
   virtual ~IDocumentStateShardHandler() = default;
-  virtual auto createLocalShard(
-      ShardID const& shardId, std::string const& collectionId,
-      std::shared_ptr<velocypack::Builder> const& properties) -> Result = 0;
-  virtual auto dropLocalShard(ShardID const& shardId,
-                              std::string const& collectionId) -> Result = 0;
+  virtual auto ensureShard(ShardID shard, CollectionID collection,
+                           std::shared_ptr<VPackBuilder> properties)
+      -> ResultT<bool> = 0;
+  virtual auto dropShard(ShardID const& shard) -> ResultT<bool> = 0;
+  virtual auto dropAllShards() -> Result = 0;
+  virtual auto isShardAvailable(ShardID const& shard) -> bool = 0;
+  virtual auto getShardMap() -> ShardMap = 0;
 };
 
 class DocumentStateShardHandler : public IDocumentStateShardHandler {
  public:
   explicit DocumentStateShardHandler(GlobalLogIdentifier gid,
                                      MaintenanceFeature& maintenanceFeature);
-  auto createLocalShard(ShardID const& shardId, std::string const& collectionId,
-                        std::shared_ptr<velocypack::Builder> const& properties)
-      -> Result override;
-  auto dropLocalShard(ShardID const& shardId, const std::string& collectionId)
-      -> Result override;
+  auto ensureShard(ShardID shard, CollectionID collection,
+                   std::shared_ptr<VPackBuilder> properties)
+      -> ResultT<bool> override;
+  auto dropShard(ShardID const& shard) -> ResultT<bool> override;
+  auto dropAllShards() -> Result override;
+  auto isShardAvailable(ShardID const& shardId) -> bool override;
+  auto getShardMap() -> ShardMap override;
+
+ private:
+  auto executeCreateCollectionAction(ShardID shard, CollectionID collection,
+                                     std::shared_ptr<VPackBuilder> properties)
+      -> Result;
+  auto executeDropCollectionAction(ShardID shard, CollectionID collection)
+      -> Result;
 
  private:
   GlobalLogIdentifier _gid;
   MaintenanceFeature& _maintenanceFeature;
+  ServerID _server;
+  struct {
```

Comment:
I wanted to use a `shared_lock` for reading data, which I didn't manage to implement with a `GuardedData`. I don't think it can differentiate between *shared* and *unique* locking.

Reply:
I created a ticket for that: [CINFRA-701](https://arangodb.atlassian.net/browse/CINFRA-701)